src/VisualStudio/Razor/RazorLanguageServiceClient.cs

```diff
@@ -22,23 +24,42 @@ internal RazorLanguageServiceClient(RemoteHostClient client, string serviceName)
             _serviceName = serviceName;
         }
 
-        public async Task<Session> CreateSessionAsync(Solution solution, object callbackTarget = null, CancellationToken cancellationToken = default)
+        public Task<Optional<T>> TryRunRemoteAsync<T>(string targetName, Solution? solution, IReadOnlyList<object?> arguments, object? callbackTarget, CancellationToken cancellationToken)
+            => _client.TryRunRemoteAsync<T>(_serviceName, targetName, solution, arguments, callbackTarget, cancellationToken);
+
+        [Obsolete("Use TryRunRemoteAsync instead")]
+        public async Task<Session?> CreateSessionAsync(Solution solution, object? callbackTarget = null, CancellationToken cancellationToken = default)
         {
             if (solution == null)
             {
                 // keep old behavior for Razor
                 return null;
             }
 
-            var innerSession = await _client.TryCreateSessionAsync(_serviceName, solution, callbackTarget, cancellationToken).ConfigureAwait(false);
-            if (innerSession == null)
+            var connection = await _client.TryCreateConnectionAsync(_serviceName, callbackTarget, cancellationToken).ConfigureAwait(false);
+            if (connection == null)
             {
                 return null;
             }
 
-            return new Session(innerSession);
+            SessionWithSolution? session = null;
+            try
+            {
+                // transfer ownership of the connection to the session object:
+                session = await SessionWithSolution.CreateAsync(connection, solution, cancellationToken).ConfigureAwait(false);
+            }
+            finally
+            {
+                if (session == null)
+                {
+                    connection.Dispose();
+                }
+            }
```

Comment:
aside: in several PRs i've seen this pattern of "create A, create B.  if B fails, cleanup A".  It would be nice to have a helper for that.

Reply:
I think we have too many helpers. I don't think this is complex enough to warrant a helper.